src/pagination.rs

```diff
@@ -0,0 +1,279 @@
+use cobalt_model::pagination_config::PaginationCfg;
+use cobalt_model::permalink;
+use cobalt_model::SortOrder;
+use document;
+use document::Document;
+use liquid;
+use std::cmp::Ordering;
+use std::ops::Range;
+use std::path::PathBuf;
+
+pub(crate) fn generate_paginators(
+  doc: &mut Document,
+  posts_data: &[liquid::Value],
+  config: &mut PaginationCfg,
+) -> Vec<liquid::Object> {
+  let mut all_posts = posts_data.to_vec();
+  match config.include.as_str() {
+    "all" => {
+      sort_posts(&mut all_posts, &config);
+      create_all_paginators(&mut all_posts, &doc, &config)
+    }
+    _ => {
+      unreachable!();
+    }
+  }
+}
+
+fn extract_page_path_from(p: &liquid::Object) -> String {
+  p.get("page_path")
+    .expect("Should have a page_path")
+    .as_scalar()
+    .expect("Should be a scalar")
+    .to_str()
+    .into_owned()
+}
+
+pub(crate) fn extract_page_path(p: &liquid::Object, config: &PaginationCfg) -> PathBuf {
+  PathBuf::from(match config.include.as_str() {
+    "all" => extract_page_path_from(&p),
+    _ => "".to_owned(),
+  })
+}
+
+fn extract_value(a: &liquid::Value, key: &String) -> Option<liquid::Scalar> {
+  a.as_object().map_or(None, |attr| {
+    attr.get(key).map_or(None, |sort_key| {
+      sort_key
+        .as_scalar()
+        .map_or(None, |value| Some(value.clone()))
+    })
+  })
+}
+
+// sort posts by multiple criteria
+fn sort_posts(posts: &mut Vec<liquid::Value>, config: &PaginationCfg) {
+  posts.sort_by(|a, b| {
+    let keys = config.sort_by.clone();
```

Comment:
Why is this clone needed?  I've also seen several others that seemed suspect

Reply:
I don't remember, but I think I was fighting the borrow checker there

Reply:
Fixed it by using a reference :)